wallet-new/src/Cardano/Wallet/API/Request/Filter.hs

```diff
@@ -63,21 +89,64 @@ data FilterOperation ix a =
     -- ^ Filter by predicate (e.g. lesser than, greater than, etc.)
     | FilterByRange ix ix
     -- ^ Filter by range, in the form [from,to]
-    | FilterIdentity
-    -- ^ Do not alter the resource.
+    deriving Eq
+
+instance ToHttpApiData ix => ToHttpApiData (FilterOperation ix a) where
+    toQueryParam = renderFilterOperation
+
+renderFilterOperation :: ToHttpApiData ix => FilterOperation ix a -> Text
+renderFilterOperation = \case
+    FilterByIndex ix ->
+        toQueryParam ix
+    FilterByPredicate p ix ->
+        mconcat [renderFilterOrdering p, "[", toQueryParam ix, "]"]
+    FilterByRange lo hi  ->
+        mconcat ["RANGE", "[", toQueryParam lo, ",", toQueryParam hi, "]"]
+
+findMatchingFilterOp
+    :: forall needle a
+    . Typeable needle
+    => FilterOperations a
+    -> Maybe (FilterOperation needle a)
+findMatchingFilterOp filters =
+    case filters of
+        NoFilters ->
+            Nothing
+        FilterOp (fop :: FilterOperation ix a) rest ->
+            case eqT @ix @needle of
+                Just Refl ->
+                    pure fop
+                Nothing ->
+                    findMatchingFilterOp rest
 
 instance Show (FilterOperation ix a) where
     show (FilterByIndex _)            = "FilterByIndex"
     show (FilterByPredicate theOrd _) = "FilterByPredicate[" <> show theOrd <> "]"
     show (FilterByRange _ _)          = "FilterByRange"
-    show FilterIdentity               = "FilterIdentity"
 
 -- | Represents a filter operation on the data model.
--- Examples:
---   *    `wallet_id=DEADBEEF`.
---   *    `balance=GT[10]`
---   *    `balance=RANGE[0,10]`
-data FilterBy (sym :: [Symbol]) (r :: *) deriving Typeable
+--
+-- The first type parameter is a type level list that pairs the query
+-- parameter string with the expected parsed type. The second type
+-- parameter describes the resource that is being filtered.
+--
+-- @
+-- 'FilterBy' '[ "id" ?= WalletId, "balance" ?= Coin ] Wallet
```

Comment:
@parsonsmatt am I dreaming or this comment, now that we scrapped `?=` is not quite right? 😁

Reply:
good catch! :smile: fixed